### PR TITLE
Add new method for custom formatter handling

### LINKED ISSUE #290: [Bug] Using multiple datasets with custom formatter fails after the first dataset.
### Describe the bug

I am training a YourTTS model using multiple internal dataset that all use the same custom formatter. I am hit two problems:
1. The embedding computing `compute_embeddings()` does not accept custom formatters as an input. Solving this seems trivial, just adding `formatter` kwarg for the method and passing it onwards to `load_tts_samples()`.
2. The `load_tts_samples()` only uses the given custom formatter for the first dataset as the formatter is set to `None` at end of each loop. And thus only the first dataset uses the custom formatter and the second one tries to use a formatter based on the formatter in the config. 

### To Reproduce

1. Define multiple datasets with custom formatter in your training script(for me the YourTTS recipe from VCTK directory). e.g. 
```
def custom_formatter():
     """Format the dataset to coqui format"""
     pass


dataset_config = BaseDatasetConfig(
    formatter='custom', meta_file_train="", language="fi-fi", path='./data/data_custom.json'
)

dataset_config_2 = BaseDatasetConfig(
    formatter='custom', meta_file_train="", language="fi-fi", path='./data/data_custom2.json' 
) 

DATASETS_CONFIG_LIST = [dataset_config, dataset_config_2]
```
2. `uv run python recipes/vctk/yourtts/train_yourtts_jackson_finetune.py` and the compute embeddings will fail with missing formatter.
3. fix `compute_embeddings()` for support custom formatter. But still the second dataset formatting fails.

### Expected behavior

_No response_

### Logs

```shell

```

### Environment

```shell
{
    "CUDA": {
        "GPU": [
            "NVIDIA GeForce RTX 4090"
        ],
        "available": true,
        "version": "12.4"
    },
    "Packages": {
        "PyTorch_debug": false,
        "PyTorch_version": "2.5.1+cu124",
        "TTS": "0.25.3",
        "numpy": "1.26.4"
    },
    "System": {
        "OS": "Linux",
        "architecture": [
            "64bit",
            "ELF"
        ],
        "processor": "x86_64",
        "python": "3.12.5",
        "version": "#98-Ubuntu SMP Mon Oct 2 15:18:56 UTC 2023"
    }
}
```

### Additional context

The problem 1. as said should be trivial and the addition of the kwarg defaulting to `None` should not cause any side effects. 

The problem 2. might require so discussion. For our usage I made a fix that if the formatter argument is given it is used for all the dataset, but I started wondering if this is actually the best solution. How this would not work for a situation where there would be a set of different formatter different formatters used. For example if you would like to fine-tune a model, that is base trained with VCTK-dataset, with a custom formatted dataset. And include some of the original VCTK-data to avoid overfitting to the new speaker.

An other way to solve this would be to change the custom dataset formatter documentation to suggest monkey patching the dataset object, but to my understanding that is not consider as a best practice in python community. Or add an api for adding and getting formatters. 

Do you love to contribute on this, but I think it having an input from others would help to reach a desired outcome. Any takes on this. Would it be easier to discuss if I open a pull request with an example fix?

### PR DESCRIPTION
**Pull request by ivuorio**
Adding a method for setting new formatter into the native formatter handling. This does not change how the existing solutions work, and thus should not cause any backwards compatibility issue. But also allows user to implement multiple custom formatters and allows using them in datasets according to their name just like the builtin formatters.  This solves both problems from the issue #290 as the the custom formatter is handled just like builtin ones.

_Adding_:
 - New method for adding custom formatters
 - Updating the documentation about the custom formatter using

_Testing_:
 - Manually tested with multiple datasets in YourTTS model training. Works with both compute embeddings and dataset loader.

_ToDo_:
- [x] Add unit tests for the new method
- [x] Add unit tests for custom formatter 
- [ ] Run full test set locally

Fixes #290

### CODE REVIEW

**Comment by eginhard** [comment]
Ah, don't try to run all tests locally, it will take forever. I've just enabled the CI for this PR, it should also rerun automatically now when you push new changes. But thanks for also adding tests! I only run them on CPU, so not sure how things go with a GPU.

I have not looked at the implementation yet and might not have time to test it in the next 2 weeks, just as a heads-up.

**Review by eginhard — APPROVED** [review]
Thanks a lot again for this PR! Really sorry it took me a long time to get to this. Everything worked well in my tests. I just added a small change to store the formatter functions in a dictionary to avoid having to manipulate the module, that seems cleaner to me.
